app/src/main/java/com/example/sharingang/utils/ImageAccess.kt

```diff
@@ -0,0 +1,188 @@
+            owner,
+            ActivityResultContracts.GetContent(),
+            ::galleryCallback
+        )
+        takePicture = registry.activityResultRegistry.register(
+            "takePicture",
+            owner,
+            ActivityResultContracts.TakePicture(),
+            ::cameraCallback
+        )
+    }
+
+    fun unregister() {
+        pickImage.unregister()
+        requestStoragePermissionLauncher.unregister()
+        takePicture.unregister()
+        requestCameraPermissionLauncher.unregister()
+    }
+
+    fun setupImageView(view: ImageView) {
+        imageView = view
+    }
+
+    private fun setupPermissionLaunchers(owner: LifecycleOwner) {
+        requestStoragePermissionLauncher = createPermissionResultRegistry(
+            "storagePermission",
+            owner,
+            Manifest.permission.READ_EXTERNAL_STORAGE
+        )
+        requestCameraPermissionLauncher =
+            createPermissionResultRegistry("cameraPermission", owner, Manifest.permission.CAMERA)
+    }
+
+    private fun createPermissionResultRegistry(
+        key: String,
+        owner: LifecycleOwner,
+        permission: String
+    ): ActivityResultLauncher<String> {
+        return registry.activityResultRegistry.register(
+            key,
+            owner,
+            ActivityResultContracts.RequestPermission()
+        ) { isGranted: Boolean ->
+            if (!isGranted)
+                Toast.makeText(
+                    registry,
+                    "$permission permission successfully denied. Feature is disabled.",
+                    Toast.LENGTH_LONG
+                ).show()
+            when (permission) {
+                Manifest.permission.CAMERA -> cameraPermissionGranted = isGranted
+                Manifest.permission.READ_EXTERNAL_STORAGE -> storagePermissionGranted = isGranted
+            }
+        }
+    }
+
+    private fun checkAndRequestPermission(
+        permission: String,
+        rationale: String,
+        context: Activity
+    ) {
+        when {
+            ContextCompat.checkSelfPermission(
+                context,
+                permission
+            ) == PackageManager.PERMISSION_GRANTED -> {
+                when (permission) {
+                    Manifest.permission.READ_EXTERNAL_STORAGE -> storagePermissionGranted = true
+                    Manifest.permission.CAMERA -> cameraPermissionGranted = true
+                }
+            }
+            shouldShowRequestPermissionRationale(context, permission) -> {
+                Toast.makeText(context, rationale, Toast.LENGTH_LONG).show()
+            }
+            else -> {
+                when (permission) {
+                    Manifest.permission.READ_EXTERNAL_STORAGE -> requestStoragePermissionLauncher.launch(
+                        permission
+                    )
+                    Manifest.permission.CAMERA -> requestCameraPermissionLauncher.launch(permission)
+                }
+            }
```

Comment:
This whole block looks similar to what is in the file PermissionRequest. Can't we refactor?

Reply:
I can't really, first because I don't have a fragment in ImageAccess (but I can surely make it take one) but most importantly because my callback isn't a ()  -> Unit, but can be a (String) -> Uri or (Uri) -> Boolean

Reply:
Fair enough